doc/design-copy-on-write-evaluation.md

```diff
@@ -0,0 +1,147 @@
+- There are opportunities for optimization, such as object pools which can nearly eliminate the allocation/initialization cost (assuming the pool is not at capacity), and streaming-stores which eliminate the read of the output buffer.
+  The overhead would thus only be a single write (for the coordinate) --- in addition to the reads for coordinate comparison.
+  We obtain for a (hypothetical) optimized implementation:
+
+  - With COW (and sharing): 2+2 reads + 2 writes.
+  - Without COW: 3+3 reads + 3 writes.
+
+  The overhead is thus 50% (*including* the overhead from coordinate comparison) for this worst-case.
+
+As for the memory overhead, the recommendation to use in-place modification holds also here and would completely eliminate the overhead.
+
+
+### 3 Complexity and other aspects
+
+In addition to memory and compute overheads there is a series of other aspects to consider.
+The brevity of the listed items is not an indicator of their importance, in many cases they may weigh *more* than what was described in the previous sections.
+
+1. Without COW the implementation of the `Dataset` library is significantly simplified.
+
+1. Without COW the design of the public-facing API is much simpler.
+   COW would lock us into something that exposes a fair bit more of the underlying implementation than we would like.
+
+1. Without COW user (developer) training is simpler, reducing the need for documentation and the teaching effort.
+
+1. Implementation of multi-threaded code using `Dataset` has a fair number of pitfalls if COW is present.
+   Without COW it becomes pretty much straightforward.
+
+1. The "worst" that can happen in a design without COW is some waste of memory.
+   With COW, developers have to be aware of invalidation of iterators and invalidation of views such as slice-views.
+   The risk of subtle bugs is thus high.
+
+1. With COW, invalidation of iterators and views and the ensuing bugs will lead to instabilities.
+   In most cases an invalid iterator will temporarily still point to valid memory, but this may change randomly in multi-threaded environments and can lead to hard-to-reproduce crashes.
+
+1. To actually get any benefit out of COW, developers need to be aware of the concept and take care not to break sharing.
+
+
+### 4 Conclusion
+
+Considered in isolation, the worst-case overheads for memory and computation are not insignificant.
+In a realistic workflow the relative importance of these overheads is likely to reduce drastically.
+
+- Not all operations/data in a workflow falls into the worst case.
+- In practice sharing may happen less often than we would like to.
+
+We can thus conjecture that the overall effect might be less than 10%, or approach zero in many cases.
+
+In comparison to the various aspects listed in Sec. 3) we may thus argue that COW is not beneficial or even counter-productive (note that we still have the freedom to use COW for scalar data, without having it as part of the library design).
```

Comment:
It is not so different:

- In C++, we either `std::move` an input to avoid copies, or provide methods instead of free functions for modifications, e.g., `Dataset::convert` instead of `Dataset convert(Dataset d)`.
- In Python we cannot `std:move`, so in-place operation is likely to be handled by methods of `Dataset`. Another option would be the numpy-style `out` arguments, but I am not a fan.